### ISSUE TITLE
error in aggregate_taxa and subsequent plot_composition

### ISSUE THREAD

**Issue by hermetia** [microbiome/microbiome#113]
I am trying to run the aggregate_taxa(pseq, "Genus") function on a phyloseq, using the microbiome package version 1.2.1. If I remove the phy_tree (phy_tree <- NULL) from the phyloseq object prior to running the function - I remember this was in a tutorial I followed previously - , I get the following error running the aggregate_taxa function:

> Error in `[<-`(`*tmp*`, nam, , value = colSums(matrix(d[taxa, ], ncol = nsamples(mypseq)))) : 
>   subscript out of bounds

When including the phy_tree, aggregate_taxa works (with the message "The top parameter to be implemented when phy_tree is available."), but plot_composition() gives an error on dimnames.

`ps.gen <- microbiome::aggregate_taxa(ps, "Genus", top = 10)`

> The top parameter to be implemented when phy_tree is available.

`plot.composition.CountAbun <- plot_composition(ps.gen, sample.sort="Description") + theme(legend.position = "bottom") + scale_fill_brewer("Genus", palette = "Paired") + theme_bw() +     theme(axis.text.x = element_text(angle = 90)) + ggtitle("Count abundance") + guide_italics + theme(legend.title = element_text(size=18))`

> Error in dimnames(x) <- dn :  length of 'dimnames' [1] not equal to array extent

Could anyone help me solve this issue? Thanks in advance.

**Comment by antagomir** [comment]
The current devel version is 1.5.31. Could you first install that with devtools::install_github("microbiome/microbiome") and then report if this problem persists?

**Comment by hermetia** [comment]
thanks, I just tried to update using your code, and it says it updated 25 packages, then I loaded them, but sessionInfo() still returns version 1.2.1 for microbiome. I use R version 3.5.0 - should I use 3.5.1? I am not so eager to update all these versions, as I've heard that R3.5.1 still had quite a few bugs and I got the advise to use an older version.

**Comment by antagomir** [comment]
If you run install_github("microbiome/microbiome") then it should install 1.5.31. Have you tried to reboot R and reload the microbiome package?

**Comment by hermetia** [comment]
yes I have tried restarting R and then reloading the package, but it still returns version 1.2.1.
libPaths() has stayed the same, so it should replace the old version by the latest upon installation, but this seems not to happen.

**Comment by antagomir** [comment]
Thanks for the report. I am not sure what would be the solution. This does not seem an issues that is specific to our package but more generic?

**Comment by hermetia** [comment]
Ok, so updating the packages in the end did not help, and only gave me lots of issues with dependencies. After restoring all to the versions I used before, I reproduced the error. Today I figured out the error was because some taxa had empty fields in the tax_table, which happens if SILVA had found too many mismatches to give the taxon a name at that level. I replaced these empty fields with the text "Unmatched_Phylum" (or instead of phylum used Class, Order, etc depending on the tax level concerned). Then I ran the code again, just like above, and all worked! So problem solved.

**Comment by hermetia** [comment]
example code for replacing empty fields with text:
`tax_table(ps)[tax_table(ps)[,"Phylum"]== "","Phylum"] <- "Unmatched_phylum"
`

**Comment by antagomir** [comment]
Thanks for the report. We might like to include this in the function in order to avoid failures. Perhaps as an optional argument because having this by default might be risky from naive user perspective.

**Comment by microsud** [comment]
This is more for a tutorial. We did this for MIB at wur. Can include this in the package website.

**Comment by antagomir** [comment]
ok. i will leave this open for now (?)

**Comment by microsud** [comment]
Yes. Will close it later after update.

**Comment by Hermandez** [comment]
Error while running the following for aggregate_top_taxa in Rstudio

> ps1.com.fam <- microbiome::aggregate_top_taxa(ps1.com, "Family", top = 10)

Error in vapply(otus, function(taxa) { : values must be type 'integer',
 but FUN(X[[1]]) result is type 'double'

**Comment by antagomir** [comment]
Could you provide a completely reproducible example (with some open phyloseq data), or send the example data for us for debugging? (or fix and make PR).

**Comment by Hermandez** [comment]
Hi @antagomir,
Here is an example of my data. I wasn't able to upload any phyloseq data due to the extension type. 
Thank you in advance for your support.  

[metadata_example.txt](https://github.com/microbiome/microbiome/files/4593182/metadata_example.txt)
[otu_table_example.txt](https://github.com/microbiome/microbiome/files/4593184/otu_table_example.txt)

[taxonomy_table_example.txt](https://github.com/microbiome/microbiome/files/4593202/taxonomy_table_example.txt)

**Comment by antagomir** [comment]
Great, thanks! Can you also provide could that loads these data files in R and reproduces the error?

**Comment by Hermandez** [comment]
I believe it can reproduce the error by following all the steps from importing the data to the phylosec then run the code. Indeed everything works perfectly until the step of plotting.  
 
  On Sat, 9 May 2020 at 1:06, Leo Lahti<notifications@github.com> wrote:   


Great, thanks! Can you also provide could that loads these data files in R and reproduces the error?

**Comment by antagomir** [comment]
Following which steps exactly. Where I can find those steps? Readily available reproducible will help in debugging.

**Comment by Hermandez** [comment]
Sorry I may have said this before. I am following this tutorial developed by you (https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/composition-plots.html)
I thought by sharing an example of my data (OTU table, taxonomy table and metadata) it could be easy to trace the error.
Perhaps I am not getting the "readily available reproducible".

**Comment by antagomir** [comment]
Could you copy-paste here the exact few lines of code that read and process the data until the error occurs?

Or send also an executable R file that does this.

I will then run those lines and see if I get the same bug. 

It is also very useful to know if there is a problem in the tutorial. But I would like to see your exact code because this speeds up things at our end, and ensure that we are talking about the same code exactly

**Comment by Hermandez** [comment]
Could this help ?


> library(microbiome) # data analysis and visualisation
Loading required package: phyloseq
Loading required package: ggplot2

microbiome R package (microbiome.github.com)
    


 Copyright (C) 2011-2020 Leo Lahti, 
    Sudarshan Shetty et al. <microbiome.github.io>


Attaching package: ‘microbiome’

The following object is masked from ‘package:ggplot2’:

    alpha

The following object is masked from ‘package:base’:

    transform

> library(phyloseq) # also the basis of data object. Data analysis and visualisation
> library(microbiomeutilities) # some utility tools
Warning message:
replacing previous import ‘ggplot2::alpha’ by ‘microbiome::alpha’ when loading ‘microbiomeutilities’ 
> library(RColorBrewer) # nice color options
> library(ggpubr) # publication quality figures, based on ggplot2
Loading required package: magrittr
> library(DT) # interactive tables in html and markdown
> library(data.table) # alternative to data.frame
data.table 1.12.8 using 2 threads (see ?getDTthreads).  Latest news: r-datatable.com
> library(dplyr) # data handling

Attaching package: ‘dplyr’

The following objects are masked from ‘package:data.table’:

    between, first, last

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union

> # Installation
> # Setup environment
> w <- getwd()
> w
[1] "C:/Users/User/Documents/Microbiome custom"
> setwd(w)
> otu <- read.table(file ="otu_table.txt", header = TRUE)
> head(otu)
                             OTUID  MK1  MK2  MC1  MC2  MM1  MM2
1 eba603cc17921993ea9d85af04a009ee 4337 1622 1092 1023 3274  102
2 5036e9ec111640d224f12e19947c11b5 4175 4894 3333 1980 8431 1045
3 a4a7c8c09d9370f3f413c4e8e07b3a66 3539 1983    4  206 2624   20
4 f93151bf0847edcab545349c2800b9ee 3518 4262  392  333 1652   30
5 45bc21dd4a57263afe36a3d8d50f1d2e 3496 1826    1  131 4112   27
6 8ec030e10ea5ab26ff4dbd9c0b0864bc 2860 2032  440  812 1831   91
   CT1
1  418
2 2570
3  269
4  156
5  486
6  215
> tax <- read.table(file="AmfSoiltaxonomy.tsv", sep='\t', header=TRUE)
> head(tax)
                             OTUID Kingdom        Phylum
1 90557237e1a619d1f5d127c6f8d55baf   Fungi Glomeromycota
2 db74c68ff315a3736ab2be6ef70d6000   Fungi Glomeromycota
3 614134b7ee3832ca4879379a8d4910ed   Fungi Glomeromycota
4 c22fa01bd6aa4463a5097daea3d5f88e   Fungi Glomeromycota
5 6257c5bd847972e7d1d2d7e6e71d9ca1   Fungi Glomeromycota
6 dbc97e5dd09488f6870582c3a6e63f10   Fungi Glomeromycota
           Class           Order          Family       Genus
1 Glomeromycetes Diversisporales Acaulosporaceae Acaulospora
2 Glomeromycetes Diversisporales Acaulosporaceae Acaulospora
3 Glomeromycetes Diversisporales Acaulosporaceae Acaulospora
4 Glomeromycetes Diversisporales Acaulosporaceae Acaulospora
5 Glomeromycetes Diversisporales Acaulosporaceae Acaulospora
6 Glomeromycetes Diversisporales Acaulosporaceae Acaulospora
                        Species
1 Acaulospora_Early-19_VTX00026
2 Acaulospora_Early-19_VTX00026
3 Acaulospora_Early-19_VTX00026
4 Acaulospora_Early-19_VTX00026
5 Acaulospora_Early-19_VTX00026
6 Acaulospora_Early-19_VTX00026
> tax_filtered <- tax[tax$OTUID %in% otu$OTUID,]
> head(tax_filtered)
                             OTUID Kingdom        Phylum
1 90557237e1a619d1f5d127c6f8d55baf   Fungi Glomeromycota
2 db74c68ff315a3736ab2be6ef70d6000   Fungi Glomeromycota
3 614134b7ee3832ca4879379a8d4910ed   Fungi Glomeromycota
4 c22fa01bd6aa4463a5097daea3d5f88e   Fungi Glomeromycota
5 6257c5bd847972e7d1d2d7e6e71d9ca1   Fungi Glomeromycota
6 dbc97e5dd09488f6870582c3a6e63f10   Fungi Glomeromycota
           Class           Order          Family       Genus
1 Glomeromycetes Diversisporales Acaulosporaceae Acaulospora
2 Glomeromycetes Diversisporales Acaulosporaceae Acaulospora
3 Glomeromycetes Diversisporales Acaulosporaceae Acaulospora
4 Glomeromycetes Diversisporales Acaulosporaceae Acaulospora
5 Glomeromycetes Diversisporales Acaulosporaceae Acaulospora
6 Glomeromycetes Diversisporales Acaulosporaceae Acaulospora
                        Species
1 Acaulospora_Early-19_VTX00026
2 Acaulospora_Early-19_VTX00026
3 Acaulospora_Early-19_VTX00026
4 Acaulospora_Early-19_VTX00026
5 Acaulospora_Early-19_VTX00026
6 Acaulospora_Early-19_VTX00026
> write.csv(tax_filtered, file="AmfSoiltaxonomy.csv")
> library("ggplot2")
> library("phyloseq")
> library("ape")

Attaching package: ‘ape’

The following object is masked from ‘package:ggpubr’:

    rotate

> otu_table <- read.csv("otu_matrix.csv", sep = ",", row.names = 1)
> otu_table <- as.matrix(otu_table)
> head(otu_table)
                                 MK1 MK2 MC1 MC2 MM1 MM2 CT1
00001016a2d5e1eea690eebeb9804fcd  25  50   0   0   0   0   0
00048ac3224d05bd66d2cd9f2a743f5b  11   9   0   0   4   0   0
000609ad63b06c8ff8caf5e7fcd5b9c0   5   7   0   0   1   0   0
0009437e8929da242c8548d7e2ffdc0d   0   0   0  46   0   0   0
000b8048c6d3b6ba359211daf2503ab1   0   0   0   0   7   3   0
000c1288bf8b8d60d77bfae4253cd870 204  61   0   0 117   0  14
> taxonomy <- read.csv("AmfSoiltaxonomy.csv", sep = ",", row.names = 1)
> taxonomy <- as.matrix(taxonomy)
> head(taxonomy)
                                 Kingdom Phylum         
90557237e1a619d1f5d127c6f8d55baf "Fungi" "Glomeromycota"
db74c68ff315a3736ab2be6ef70d6000 "Fungi" "Glomeromycota"
614134b7ee3832ca4879379a8d4910ed "Fungi" "Glomeromycota"
c22fa01bd6aa4463a5097daea3d5f88e "Fungi" "Glomeromycota"
6257c5bd847972e7d1d2d7e6e71d9ca1 "Fungi" "Glomeromycota"
dbc97e5dd09488f6870582c3a6e63f10 "Fungi" "Glomeromycota"
                                 Class            Order            
90557237e1a619d1f5d127c6f8d55baf "Glomeromycetes" "Diversisporales"
db74c68ff315a3736ab2be6ef70d6000 "Glomeromycetes" "Diversisporales"
614134b7ee3832ca4879379a8d4910ed "Glomeromycetes" "Diversisporales"
c22fa01bd6aa4463a5097daea3d5f88e "Glomeromycetes" "Diversisporales"
6257c5bd847972e7d1d2d7e6e71d9ca1 "Glomeromycetes" "Diversisporales"
dbc97e5dd09488f6870582c3a6e63f10 "Glomeromycetes" "Diversisporales"
                                 Family            Genus        
90557237e1a619d1f5d127c6f8d55baf "Acaulosporaceae" "Acaulospora"
db74c68ff315a3736ab2be6ef70d6000 "Acaulosporaceae" "Acaulospora"
614134b7ee3832ca4879379a8d4910ed "Acaulosporaceae" "Acaulospora"
c22fa01bd6aa4463a5097daea3d5f88e "Acaulosporaceae" "Acaulospora"
6257c5bd847972e7d1d2d7e6e71d9ca1 "Acaulosporaceae" "Acaulospora"
dbc97e5dd09488f6870582c3a6e63f10 "Acaulosporaceae" "Acaulospora"
                                 Species                        
90557237e1a619d1f5d127c6f8d55baf "Acaulospora_Early-19_VTX00026"
db74c68ff315a3736ab2be6ef70d6000 "Acaulospora_Early-19_VTX00026"
614134b7ee3832ca4879379a8d4910ed "Acaulospora_Early-19_VTX00026"
c22fa01bd6aa4463a5097daea3d5f88e "Acaulospora_Early-19_VTX00026"
6257c5bd847972e7d1d2d7e6e71d9ca1 "Acaulospora_Early-19_VTX00026"
dbc97e5dd09488f6870582c3a6e63f10 "Acaulospora_Early-19_VTX00026"
> metadata <- read.table("metadata.txt", row.names = 1, header = T) 
> metadata
    BarcodeSequencei5 BarcodeSequencei7 Index_i5 Index_i7 PrimerSet
MC1          TATCCTCT          CAGAGAGG     S503     N708 S503_N708
MC2          TATCCTCT          GCTACGCT     S504     N709 S503_N709
MM1          TATCCTCT          TCCTGAGC     S505     N704 S503_N704
MM2          TATCCTCT          AAGAGGCA     S506     N711 S503_N711
CT1          TATCCTCT          GTAGAGGA     S507     N712 S503_N712
MK1          TATCCTCT          GGACTCCT     S509     N705 S503_N705
MK2          TATCCTCT          CTCTCTAC     S510     N707 S503_N707
       Sample_description   Location                     Forward
MC1 fungiDNAPafricanaSoil    MntCmr1  S2_S1_L001_R1_001_fastq.gz
MC2 fungiDNAPafricanaSoil    MntCmr2  S3_S2_L001_R1_001_fastq.gz
MM1 fungiDNAPafricanaSoil   MntMgba1  S4_S3_L001_R1_001_fastq.gz
MM2 fungiDNAPafricanaSoil   MntMgba2  S6_S4_L001_R1_001_fastq.gz
CT1 fungiDNAPafricanaSoil CkaTNithi1  S7_S5_L001_R1_001_fastq.gz
MK1 fungiDNAPafricanaSoil  MlvaKkga1 S10_S7_L001_R1_001_fastq.gz
MK2 fungiDNAPafricanaSoil  MlvaKkga2 S12_S8_L001_R1_001_fastq.gz
                        Reverse Provenance  pH     P    Na      K
MC1  S2_S1_L001_R2_001_fastq.gz   Cameroon 6.7 37.33 81.49 255.31
MC2  S3_S2_L001_R2_001_fastq.gz   Cameroon 6.7 37.33 81.49 255.31
MM1  S4_S3_L001_R2_001_fastq.gz   Cameroon 6.2 35.14 76.68 565.09
MM2  S6_S4_L001_R2_001_fastq.gz   Cameroon 6.2 35.14 76.68 565.09
CT1  S7_S5_L001_R2_001_fastq.gz      Kenya 5.9  6.04 11.66 555.18
MK1 S10_S7_L001_R2_001_fastq.gz      Kenya 6.8  7.93 18.25 564.02
MK2 S12_S8_L001_R2_001_fastq.gz      Kenya 6.8  7.93 18.25 564.02
> # Read in tree
> phy_tree <- read_tree("tree.nwk")
> OTU <- otu_table(otu_table, taxa_are_rows = TRUE)
> TAX <- tax_table(taxonomy)
> META <- sample_data(metadata)
> head(TAX)
Taxonomy Table:     [6 taxa by 7 taxonomic ranks]:
                                 Kingdom Phylum         
90557237e1a619d1f5d127c6f8d55baf "Fungi" "Glomeromycota"
db74c68ff315a3736ab2be6ef70d6000 "Fungi" "Glomeromycota"
614134b7ee3832ca4879379a8d4910ed "Fungi" "Glomeromycota"
c22fa01bd6aa4463a5097daea3d5f88e "Fungi" "Glomeromycota"
6257c5bd847972e7d1d2d7e6e71d9ca1 "Fungi" "Glomeromycota"
dbc97e5dd09488f6870582c3a6e63f10 "Fungi" "Glomeromycota"
                                 Class            Order            
90557237e1a619d1f5d127c6f8d55baf "Glomeromycetes" "Diversisporales"
db74c68ff315a3736ab2be6ef70d6000 "Glomeromycetes" "Diversisporales"
614134b7ee3832ca4879379a8d4910ed "Glomeromycetes" "Diversisporales"
c22fa01bd6aa4463a5097daea3d5f88e "Glomeromycetes" "Diversisporales"
6257c5bd847972e7d1d2d7e6e71d9ca1 "Glomeromycetes" "Diversisporales"
dbc97e5dd09488f6870582c3a6e63f10 "Glomeromycetes" "Diversisporales"
                                 Family            Genus        
90557237e1a619d1f5d127c6f8d55baf "Acaulosporaceae" "Acaulospora"
db74c68ff315a3736ab2be6ef70d6000 "Acaulosporaceae" "Acaulospora"
614134b7ee3832ca4879379a8d4910ed "Acaulosporaceae" "Acaulospora"
c22fa01bd6aa4463a5097daea3d5f88e "Acaulosporaceae" "Acaulospora"
6257c5bd847972e7d1d2d7e6e71d9ca1 "Acaulosporaceae" "Acaulospora"
dbc97e5dd09488f6870582c3a6e63f10 "Acaulosporaceae" "Acaulospora"
                                 Species                        
90557237e1a619d1f5d127c6f8d55baf "Acaulospora_Early-19_VTX00026"
db74c68ff315a3736ab2be6ef70d6000 "Acaulospora_Early-19_VTX00026"
614134b7ee3832ca4879379a8d4910ed "Acaulospora_Early-19_VTX00026"
c22fa01bd6aa4463a5097daea3d5f88e "Acaulospora_Early-19_VTX00026"
6257c5bd847972e7d1d2d7e6e71d9ca1 "Acaulospora_Early-19_VTX00026"
dbc97e5dd09488f6870582c3a6e63f10 "Acaulospora_Early-19_VTX00026"
> tail(TAX)
Taxonomy Table:     [6 taxa by 7 taxonomic ranks]:
                                 Kingdom Phylum         
2be143735b6f69be74c5308bf8997b68 "Fungi" "Glomeromycota"
f867f9494565484597083ff1b7e6002d "Fungi" "Glomeromycota"
80723c6b55fd65b15099596619528024 "Fungi" "Glomeromycota"
b1735395f360859b34c7fdd001fe5b68 "Fungi" "Glomeromycota"
7bda2d4fe61e1cd9a23a627f35e59841 "Fungi" "Glomeromycota"
a7777059958b61094f1bf8349f2bdef2 "Fungi" "Glomeromycota"
                                 Class            Order       
2be143735b6f69be74c5308bf8997b68 "Glomeromycetes" "Glomerales"
f867f9494565484597083ff1b7e6002d "Glomeromycetes" "Glomerales"
80723c6b55fd65b15099596619528024 "Glomeromycetes" "Glomerales"
b1735395f360859b34c7fdd001fe5b68 "Glomeromycetes" "Glomerales"
7bda2d4fe61e1cd9a23a627f35e59841 "Glomeromycetes" "Glomerales"
a7777059958b61094f1bf8349f2bdef2 "Glomeromycetes" "Glomerales"
                                 Family        Genus Species
2be143735b6f69be74c5308bf8997b68 "Glomeraceae" ""    ""     
f867f9494565484597083ff1b7e6002d "Glomeraceae" ""    ""     
80723c6b55fd65b15099596619528024 "Glomeraceae" ""    ""     
b1735395f360859b34c7fdd001fe5b68 "Glomeraceae" ""    ""     
7bda2d4fe61e1cd9a23a627f35e59841 "Glomeraceae" ""    ""     
a7777059958b61094f1bf8349f2bdef2 "Glomeraceae" ""    ""     
> head(OTU)
OTU Table:          [6 taxa and 7 samples]
                     taxa are rows
                                 MK1 MK2 MC1 MC2 MM1 MM2 CT1
00001016a2d5e1eea690eebeb9804fcd  25  50   0   0   0   0   0
00048ac3224d05bd66d2cd9f2a743f5b  11   9   0   0   4   0   0
000609ad63b06c8ff8caf5e7fcd5b9c0   5   7   0   0   1   0   0
0009437e8929da242c8548d7e2ffdc0d   0   0   0  46   0   0   0
000b8048c6d3b6ba359211daf2503ab1   0   0   0   0   7   3   0
000c1288bf8b8d60d77bfae4253cd870 204  61   0   0 117   0  14
> taxa_names(TAX)[1:10]
 [1] "90557237e1a619d1f5d127c6f8d55baf"
 [2] "db74c68ff315a3736ab2be6ef70d6000"
 [3] "614134b7ee3832ca4879379a8d4910ed"
 [4] "c22fa01bd6aa4463a5097daea3d5f88e"
 [5] "6257c5bd847972e7d1d2d7e6e71d9ca1"
 [6] "dbc97e5dd09488f6870582c3a6e63f10"
 [7] "b183019b1c964a107dd3b9e4c63d6527"
 [8] "4058bc8d42a9fc41e2382c94ffcd39b1"
 [9] "7b69ed442f91019ed3c63bbe8a6ede74"
[10] "8ce4414ced1b0f7411fc18da832d7d24"
> taxa_names(OTU)[1:10]
 [1] "00001016a2d5e1eea690eebeb9804fcd"
 [2] "00048ac3224d05bd66d2cd9f2a743f5b"
 [3] "000609ad63b06c8ff8caf5e7fcd5b9c0"
 [4] "0009437e8929da242c8548d7e2ffdc0d"
 [5] "000b8048c6d3b6ba359211daf2503ab1"
 [6] "000c1288bf8b8d60d77bfae4253cd870"
 [7] "000e7fd683e4151f3d416982b773e99e"
 [8] "000ef8bf740e50e7e052fed91be62ae4"
 [9] "000ff9caa038052ac947ed8774dc4ed3"
[10] "0010ad2bf3c3e383d5b9468fa01e0be6"
> taxa_names(phy_tree)[1:10]
 [1] "d04fff6823e07e1caa8a6663ed4a5dbb"
 [2] "c5a3d131d309b9b71a1896d7377179b8"
 [3] "8235c6c7156c339c18d68caec18c1349"
 [4] "944100bb1b79b5f263766f1603eafd4f"
 [5] "9f3994c7eb6dd62c39ae9430d31bef8d"
 [6] "3e7b81f8d6c5130d0320e307fe41138e"
 [7] "33d8a932e396e5ca4fee384455a5b7cd"
 [8] "86825b3e61e13b7d2bfd4cb8fa149bc8"
 [9] "132ce1a74b2434d8c5f87865ca8107db"
[10] "ddb357f42d55b304650e4ec201976c0c"
> sample_names(OTU)
[1] "MK1" "MK2" "MC1" "MC2" "MM1" "MM2" "CT1"
> sample_names(META)
[1] "MC1" "MC2" "MM1" "MM2" "CT1" "MK1" "MK2"
> Yves_phy <- phyloseq(OTU, TAX, META, phy_tree)
> Yves_phy
phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 16279 taxa and 7 samples ]
sample_data() Sample Data:       [ 7 samples by 14 sample variables ]
tax_table()   Taxonomy Table:    [ 16279 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 16279 tips and 16269 internal nodes ]
> print(Yves_phy)
phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 16279 taxa and 7 samples ]
sample_data() Sample Data:       [ 7 samples by 14 sample variables ]
tax_table()   Taxonomy Table:    [ 16279 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 16279 tips and 16269 internal nodes ]
> rank_names(Yves_phy) # we check the taxonomic rank information
[1] "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"  
[7] "Species"
> datatable(tax_table(Yves_phy)) # the table is interactive you can scrol and search thorugh it for details.
Warning message:
In instance$preRenderHook(instance) :
  It seems your data is too big for client-side DataTables. You may consider server-side processing: https://rstudio.github.io/DT/server.html
> tax_table(Yves_phy)[tax_table(Yves_phy)[, "Phylum"] == "", "Phylum"] <- "Unidentified"
> saveRDS(Yves_phy, "./R_downstream/Yves_phy.rds")
> Yves_phy <- readRDS("./R_downstream/Yves_phy.rds")
> # check for features of data  
> summarize_phyloseq(Yves_phy)
Compositional = NO2
1] Min. number of reads = 671362] Max. number of reads = 5047153] Total number of reads = 17859544] Average number of reads = 255136.2857142865] Median number of reads = 2480027] Sparsity = 0.6526287153475566] Any OTU sum to 1 or less? NO8] Number of singletons = 09] Percent of OTUs that are singletons 
        (i.e. exactly one read detected across all samples)010] Number of sample variables are: 14BarcodeSequencei5BarcodeSequencei7Index_i5Index_i7PrimerSetSample_descriptionLocationForwardReverseProvenancepHPNaK2
[[1]]
[1] "1] Min. number of reads = 67136"

[[2]]
[1] "2] Max. number of reads = 504715"

[[3]]
[1] "3] Total number of reads = 1785954"

[[4]]
[1] "4] Average number of reads = 255136.285714286"

[[5]]
[1] "5] Median number of reads = 248002"

[[6]]
[1] "7] Sparsity = 0.652628715347556"

[[7]]
[1] "6] Any OTU sum to 1 or less? NO"

[[8]]
[1] "8] Number of singletons = 0"

[[9]]
[1] "9] Percent of OTUs that are singletons \n        (i.e. exactly one read detected across all samples)0"

[[10]]
[1] "10] Number of sample variables are: 14"

[[11]]
 [1] "BarcodeSequencei5"  "BarcodeSequencei7"  "Index_i5"          
 [4] "Index_i7"           "PrimerSet"          "Sample_description"
 [7] "Location"           "Forward"            "Reverse"           
[10] "Provenance"         "pH"                 "P"                 
[13] "Na"                 "K"                 

> Yves_phy1 <- subset_taxa(Yves_phy, Family != "Glomeraceae")
> ntaxa(Yves_phy)-ntaxa(Yves_phy1)
[1] 6068
> #total number of reads in the dataset
> reads_per_asv1 <- taxa_sums(Yves_phy1)
> print(sum(reads_per_asv1))
[1] 1204532
> #How many ASVs are less than 10 reads and how many reads do they contain?
> print(length(reads_per_asv1[reads_per_asv1 < 10]))
[1] 0
> ps1 <- readRDS("./R_downstream/Yves_phy.rds")
> print(ps1)
phyloseq-class experiment-level object
otu_table()   OTU Table:         [ 16279 taxa and 7 samples ]
sample_data() Sample Data:       [ 7 samples by 14 sample variables ]
tax_table()   Taxonomy Table:    [ 16279 taxa by 7 taxonomic ranks ]
phy_tree()    Phylogenetic Tree: [ 16279 tips and 16269 internal nodes ]
> ps1.com <- ps1
> # if you have dada2/deblur output and sequences as taxa names, then you can change them as follows
> taxa_names(ps1.com) <- paste0("ASV_", rownames(tax_table(ps1.com)))
> # We need to set Palette
> taxic <- as.data.frame(ps1.com@tax_table) # this will help in setting large color options
> taxic$OTU <- rownames(taxic) # Add the OTU ids from OTU table into the taxa table at the end.
> colnames(taxic) # You can see that we now have extra taxonomy levels.
[1] "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"  
[7] "Species" "OTU"    
> taxmat <- as.matrix(taxic) # convert it into a matrix.
> new.tax <- tax_table(taxmat) # convert into phyloseq compatible file.
> tax_table(ps1.com) <- new.tax # incroporate into phyloseq Object
> # now edit the unclassified taxa
> tax_table(ps1.com)[tax_table(ps1.com)[, "Family"] == "", "Family"] <- "Unclassified family"
> guide_italics <- guides(fill = guide_legend(label.theme = element_text(
+   size = 15,
+   face = "italic", colour = "Black", angle = 0
+ )))
> ps1.com@phy_tree <- NULL
**> ps1.com.fam <- microbiome::aggregate_top_taxa(ps1.com, "Family", top = 10)
Error in vapply(otus, function(taxa) { : values must be type 'integer',
 but FUN(X[[1]]) result is type 'double'**
>

**Comment by antagomir** [comment]
No. I only mean the R code that I can run on my own computer. I do not mean the output of that code. If I copy paste the text from above, it is not executable as such because it contains also other material than source code (i.e. output of the code). Can you just send the R code that you use to read the data and run the analyses? You can send the code by copypasting the code here, or by sending the R file that you have run on your own computer.

This is the typical procedure for handling this kind of issues. The user provides code that the developer can run and try to fix.

**Comment by Hermandez** [comment]
Okay. 
I will do it right now

**Comment by Hermandez** [comment]
Where is the R code Using the tutorial


library("microbiome") # data analysis and visualisation
library("phyloseq") # also the basis of data object. Data analysis and visualisation
library("microbiomeutilities") # some utility tools
library("RColorBrewer") # nice color options
library("ggpubr") # publication quality figures, based on ggplot2
library("DT") # interactive tables in html and markdown
library("data.table") # alternative to data.frame
library("dplyr") # data handling
library("ape")

# Installation
# Setup environment

w <- getwd()
w
setwd(w)

dir()


# read in OTU table 

otu <- read.table(file ="otu_table_exple.txt", header = TRUE)
head(otu)

#Read in the .tsv version of the taxonomy table, which should also have a column header "OTUID", not "Feature ID"

tax <- read.table(file="taxonomy_table_exple.tsv", sep='\t', header=TRUE)
head(tax)

#Create a list of OTUIDs that are present in tax, but not in otus, which need to be eliminated and remove them
#The remaining table should have the same number of rows as in the otus data frame

tax_filtered <- tax[tax$OTUID %in% otu$OTUID,]
head(tax_filtered)


#Separate the "Taxon" column in the tax_filtered data frame by semicolon so that each step of the taxonomy (e.g., kingdom, phylum, class, etc.) is its own column

#tax_filtered <- separate(tax_filtered, Taxon, c("Kingdom","Phylum","Class","Order", "Family","Species"), sep= ";", remove=TRUE)

#write one outfile containing the OTUID and taxonomic info

write.csv(tax_filtered, file="taxonomy_table_exple.csv")

# Read in OTU table

otu_table <- read.csv("otu_matrix_exple.csv", sep = ",", row.names = 1)
otu_table <- as.matrix(otu_table)
head(otu_table)
# Read in taxonomy
# Separated by kingdom, phylum, class, order, family, genus, species

taxonomy <- read.csv("taxonomy_table_exple.csv", sep = ",", row.names = 1)
taxonomy <- as.matrix(taxonomy)
head(taxonomy)

# Read in metadata
# variables = SampleID, BarcodeSequencei5, BarcodeSequencei7, Index_i5, Index_i7, PrimerSet
# variables = Sample_description, Sample_collection_site, 

metadata <- read.table("metadata_exple.txt", row.names = 1, header = T) 
metadata



# Read in tree
#phy_tree <- read_tree("tree.nwk")

# Import all as phyloseq objects

OTU <- otu_table(otu_table, taxa_are_rows = TRUE)
TAX <- tax_table(taxonomy)
META <- sample_data(metadata)
head(TAX)
tail(TAX)
head(OTU)

# Sanity checks for consistent OTU names

taxa_names(TAX)[1:10]
taxa_names(OTU)[1:10]
#taxa_names(phy_tree)[1:10]

# make sure files have the same sample names

sample_names(OTU)
sample_names(META)

# Finally merge everething into one phyloseq object

Yves_phy <- phyloseq(OTU, TAX, META)
Yves_phy

#Composition plots

ps1 <- Yves_phy

# use print option to see the data saved as phyloseq object.

print(ps1)

ps1.com <- ps1

# if you have dada2/deblur output and sequences as taxa names, then you can change them as follows
taxa_names(ps1.com) <- paste0("ASV_", rownames(tax_table(ps1.com)))

# We need to set Palette
taxic <- as.data.frame(ps1.com@tax_table) # this will help in setting large color options

# colourCount = length(unique(taxic$Family))  #define number of variable colors based on number of Family (change the level accordingly to phylum/class/order)
# getPalette = colorRampPalette(brewer.pal(12, "Paired"))  # change the palette as well as the number of colors will change according to palette.

taxic$OTU <- rownames(taxic) # Add the OTU ids from OTU table into the taxa table at the end.
colnames(taxic) # You can see that we now have extra taxonomy levels.


taxmat <- as.matrix(taxic) # convert it into a matrix.
new.tax <- tax_table(taxmat) # convert into phyloseq compatible file.
tax_table(ps1.com) <- new.tax # incroporate into phyloseq Object


# now edit the unclassified taxa
tax_table(ps1.com)[tax_table(ps1.com)[, "Family"] == "", "Family"] <- "Unclassified family"

# it would be nice to have the Taxonomic names in italics.
# for that we set this

guide_italics <- guides(fill = guide_legend(label.theme = element_text(
  size = 15,
  face = "italic", colour = "Black", angle = 0
)))


## Now we need to plot at family level, we can do it as follows:

# first remove the phy_tree

ps1.com@phy_tree <- NULL

# Second merge at family level

ps1.com.fam <- microbiome::aggregate_top_taxa(ps1.com, "Family", top = 10)

plot.composition.COuntAbun <- plot_composition(ps1.com.fam) + theme(legend.position = "bottom") +
  scale_fill_brewer("Family", palette = "Paired") + theme_bw() +
  theme(axis.text.x = element_text(angle = 90)) +
  ggtitle("Relative abundance") + guide_italics + theme(legend.title = element_text(size = 18))

plot.composition.COuntAbun
[metadata_exple.txt](https://github.com/microbiome/microbiome/files/4602253/metadata_exple.txt)
[otu_matrix_exple.txt] (https://github.com/microbiome/microbiome/files/4602255/otu_matrix_exple.txt)
[otu_table_exple.txt](https://github.com/microbiome/microbiome/files/4602256/otu_table_exple.txt)
[taxonomy_table_exple.txt](https://github.com/microbiome/microbiome/files/4602257/taxonomy_table_exple.txt)

otu_matrix_exple.txt should be converted into otu_matrix_expel.csv
taxonomy_table_exple.txt should be converted into taxonomy_table_expel.csv 
For the code to run

Thank you so much for your assistance

**Comment by Hermandez** [comment]
library("microbiome") 
library("phyloseq") 
library("microbiomeutilities") 
library("RColorBrewer") 
library("ggpubr") 
library("DT") 
library("data.table") 
library("dplyr") 
library("ape")

otu <- read.table(file ="otu_table_exple.txt", header = TRUE)
head(otu)

tax <- read.table(file="taxonomy_table_exple.tsv", sep='\t', header=TRUE)
head(tax)

tax_filtered <- tax[tax$OTUID %in% otu$OTUID,]
head(tax_filtered)

write.csv(tax_filtered, file="taxonomy_table_exple.csv")


otu_table <- read.csv("otu_matrix_exple.csv", sep = ",", row.names = 1)
otu_table <- as.matrix(otu_table)
head(otu_table)
taxonomy <- read.csv("taxonomy_table_exple.csv", sep = ",", row.names = 1)
taxonomy <- as.matrix(taxonomy)
head(taxonomy)
metadata <- read.table("metadata_exple.txt", row.names = 1, header = T) 
metadata
OTU <- otu_table(otu_table, taxa_are_rows = TRUE)
TAX <- tax_table(taxonomy)
META <- sample_data(metadata)
head(TAX)
tail(TAX)
head(OTU)

taxa_names(TAX)[1:10]
taxa_names(OTU)[1:10]
sample_names(OTU)
sample_names(META)
Yves_phy <- phyloseq(OTU, TAX, META)
Yves_phy

ps1 <- Yves_phy

print(ps1)

ps1.com <- ps1

taxa_names(ps1.com) <- paste0("ASV_", rownames(tax_table(ps1.com)))

taxic <- as.data.frame(ps1.com@tax_table)
taxic$OTU <- rownames(taxic) 
colnames(taxic)

taxmat <- as.matrix(taxic) 
new.tax <- tax_table(taxmat)
tax_table(ps1.com) <- new.tax

tax_table(ps1.com)[tax_table(ps1.com)[, "Family"] == "", "Family"] <- "Unclassified family"

guide_italics <- guides(fill = guide_legend(label.theme = element_text(
  size = 15,
  face = "italic", colour = "Black", angle = 0
)))



ps1.com@phy_tree <- NULL


ps1.com.fam <- microbiome::aggregate_top_taxa(ps1.com, "Family", top = 10)

plot.composition.COuntAbun <- plot_composition(ps1.com.fam) + theme(legend.position = "bottom") +
  scale_fill_brewer("Family", palette = "Paired") + theme_bw() +
  theme(axis.text.x = element_text(angle = 90)) +
  ggtitle("Relative abundance") + guide_italics + theme(legend.title = element_text(size = 18))

plot.composition.COuntAbun

**Comment by antagomir** [comment]
This code fails with the following error:

`otu <- read.table(file ="otu_table_exple.txt", header = TRUE)`
 cannot open file 'otu_table_exple.txt': No such file or directory

This seems to be related to the fact that the code has a typo in the file name (otu_table_exple.txt vs otu_table_example.txt). Have you tested yourself that this exact code runs through based on the data files (and exact names of those files) that you sent above? 

Kindly provide source code that is fully tested with those example data files, so that we can focus on fixing the actual reported problem instead of this code.

For debugging it would be very useful to provide a _minimal example_ which means that all lines that are absolutely necessary in the code to reproduce the reported error, are removed. For instance, it seems that the various "head", "write.csv", "ggplot" etc commands are not related to the reported bug.

These are standard procedures in collaborative development and bug reporting.

**Comment by Hermandez** [comment]
##The following code works for me with the file attached. I hope it does for you.

library("microbiome") 
library("phyloseq") 
library("microbiomeutilities") 
library("RColorBrewer") 
library("ggpubr") 
library("DT")
library("data.table") 
library("dplyr") 
library("ape")

otu <- read.table(file ="otu_table_exple.txt", header = TRUE)
head(otu)

tax <- read.table(file="taxonomy_table_exple.tsv", sep='\t', header=TRUE)
head(tax)

tax_filtered <- tax[tax$OTUID %in% otu$OTUID,]
head(tax_filtered)

write.csv(tax_filtered, file="taxonomy_table_exple.csv")


otu_table <- read.csv("otu_matrix_exple.csv", sep = ",", row.names = 1)
otu_table <- as.matrix(otu_table)
head(otu_table)
taxonomy <- read.csv("taxonomy_table_exple.csv", sep = ",", row.names = 1)
taxonomy <- as.matrix(taxonomy)
head(taxonomy)
metadata <- read.table("metadata_exple.txt", row.names = 1, header = T) 
metadata
OTU <- otu_table(otu_table, taxa_are_rows = TRUE)
TAX <- tax_table(taxonomy)
META <- sample_data(metadata)
head(TAX)
tail(TAX)
head(OTU)

taxa_names(TAX)[1:10]
taxa_names(OTU)[1:10]
sample_names(OTU)
sample_names(META)
Yves_phy <- phyloseq(OTU, TAX, META)
Yves_phy

ps1 <- Yves_phy

print(ps1)

ps1.com <- ps1

taxa_names(ps1.com) <- paste0("ASV_", rownames(tax_table(ps1.com)))

taxic <- as.data.frame(ps1.com@tax_table)
taxic$OTU <- rownames(taxic) 
colnames(taxic)

taxmat <- as.matrix(taxic) 
new.tax <- tax_table(taxmat)
tax_table(ps1.com) <- new.tax

tax_table(ps1.com)[tax_table(ps1.com)[, "Family"] == "", "Family"] <- "Unclassified family"

guide_italics <- guides(fill = guide_legend(label.theme = element_text(
  size = 15,
  face = "italic", colour = "Black", angle = 0
)))


ps1.com@phy_tree <- NULL


ps1.com.fam <- microbiome::aggregate_top_taxa(ps1.com, "Family", top = 10)

plot.composition.COuntAbun <- plot_composition(ps1.com.fam) + theme(legend.position = "bottom") +
  scale_fill_brewer("Family", palette = "Paired") + theme_bw() +
  theme(axis.text.x = element_text(angle = 90)) +
  ggtitle("Relative abundance") + guide_italics + theme(legend.title = element_text(size = 18))

plot.composition.COuntAbun

[metadata_exple.txt](https://github.com/microbiome/microbiome/files/4604209/metadata_exple.txt)
[otu_table_exple.txt](https://github.com/microbiome/microbiome/files/4604210/otu_table_exple.txt)
[taxonomy_table_exple.txt](https://github.com/microbiome/microbiome/files/4604211/taxonomy_table_exple.txt)

Kindly change the extension of the file **taxonomy_table_expel** from **txt** to **tsv** before running the code.

**Comment by antagomir** [comment]
Thanks. 

This code fails with:

Error in file(file, "rt") : cannot open the connection
In addition: Warning message:
In file(file, "rt") :
  cannot open file 'otu_matrix_exple.csv': No such file or directory

I also see a lot of plotting code that is not related to the bug that we are discussing.

Can you do the following in order to facilitate debugging:

1) Remove all extra code that is not related to the reported bug

2) Provide a zip file that includes example data files and the source code that reproduces the bug

3) Test that the code in the zip file runs without error in R with source("myfile.R")

4) Once this minimal example is ready, send it to us and we test if the code is executable and reproduces the bug. If yes, we can start debugging.

**Comment by Hermandez** [comment]
Hi Leo,
I hope this will work now. 
Sorry for not getting you in the first place.
I never went through this process before.  

[Exple_data.zip](https://github.com/microbiome/microbiome/files/4607853/Exple_data.zip)

**Comment by Hermandez** [comment]
![image](https://user-images.githubusercontent.com/50687137/81553356-9fddb380-938d-11ea-8e46-29712c10000c.png)
[file_show.pdf](https://github.com/microbiome/microbiome/files/4609080/file_show.pdf)

Hello Leo,

I think the bugs are due to the R version. (I am currently using **R 4.0.0**) 
I tried to run the code on  Rstudio cloud (using **R 3.6.0**) and it works. I have attached the image of the plots I got from the code.

**Comment by antagomir** [comment]
Should work now with the latest devel version.

Install with
devtools::install_github("microbiome/microbiome")

**Comment by Hermandez** [comment]
Thank you very much
